File: app/src/main/java/com/przemyslawlusnia/vocabularycreator/wordlist/domain/AddWordUseCase.java
package com.przemyslawlusnia.vocabularycreator.wordlist.domain;

import android.support.annotation.NonNull;
import com.przemyslawlusnia.vocabularycreator.core.RxUseCase;
import com.przemyslawlusnia.vocabularycreator.wordlist.repository.WordsRepository;
import rx.Observable;
import rx.Scheduler;

public class AddWordUseCase extends RxUseCase<Boolean> {

  private WordDomainModel wordDomainModel;

  private final WordsRepository wordsRepository;

  public AddWordUseCase(Scheduler executionScheduler, WordsRepository wordsRepository) {
    super(executionScheduler);
    this.wordsRepository = wordsRepository;
  }

  public AddWordUseCase init(@NonNull WordDomainModel wordDomainModel) {
    this.wordDomainModel = wordDomainModel;
    return this;
  }

  @Override
  protected Observable<Boolean> buildUseCaseObservable() {
    if (this.wordDomainModel == null) {
      throw new IllegalArgumentException("init(wordDomainModel) not called, or called with null argument.");
    }
    return Observable.concat(validate(), wordsRepository.add(wordDomainModel));
  }

  private Observable<Boolean> validate() {
    return Observable.create(subscriber -> {
      if (AddWordUseCase.this.wordDomainModel.getWord().isEmpty()) {
        subscriber.onError(new Throwable("Error: word cannot be empty."));
      } else {
        subscriber.onCompleted();
      }
    });
  }
}

File: app/src/main/java/com/przemyslawlusnia/vocabularycreator/wordlist/domain/AddWordUseCase.kt
package com.przemyslawlusnia.vocabularycreator.wordlist.domain

import com.przemyslawlusnia.vocabularycreator.core.RxUseCase
import com.przemyslawlusnia.vocabularycreator.wordlist.repository.WordsRepository
import rx.Observable
import rx.Scheduler

class AddWordUseCase(executionScheduler: Scheduler, private val wordsRepository: WordsRepository)
    : RxUseCase<Boolean>(executionScheduler) {

    lateinit private var wordDomainModel: WordDomainModel

    fun initialize(wordDomainModel: WordDomainModel) {
        this.wordDomainModel = wordDomainModel
    }

    public override fun buildUseCaseObservable(): Observable<Boolean> {
        return Observable.concat(validate(), wordsRepository.add(wordDomainModel))
    }

    private fun validate(): Observable<Boolean> {
        return Observable.create<Boolean> { subscriber ->
            if (wordDomainModel.word.isEmpty()) {
                subscriber.onError(Throwable("Error: word cannot be empty."))
            } else {
                subscriber.onCompleted()
            }
        }
    }
}
